File: sources/gpcp/java/MsilAsm.java
// (* ========================================================= *)
// (**	Interface to the ILASM Byte-code assembler.		*)
// (*	K John Gough, 10th June 1999				*)
// (*	Modifications:						*)
// (*		Version for GPCP V0.3 April 2000 (kjg)		*)
// (* ========================================================= *)
// (*	The real code is in MsilAsm.cs or MsilAsm.java          *)
// (* ========================================================= *)
//
//MODULE MsilAsm;
//
//  PROCEDURE Init*(); BEGIN END Init;
//
//  PROCEDURE Assemble*(IN fil,opt : ARRAY OF CHAR; main : BOOLEAN); 
//  BEGIN END Assemble;
//
//  PROCEDURE DoAsm*(IN fil,opt : ARRAY OF CHAR; 
//                    main,vbse : BOOLEAN;
//                     OUT rslt : INTEGER); 
//  BEGIN END Assemble;
//
//END MsilAsm.
// 
//  NOTES:
//    This code assumes that ilasm.exe is visible on the path.
//    If the DoAsm call fails with a "cannot find ilasm" error
//    then you will have to locate it in the file system.
//    The program exists if you have the .NET JDK (instead of
//    just the runtime), or if you have any recent Visual Studio
//    releases. 
//    On Windows try searching C:\Windows\Microsoft.NET\Framework
//    and choose the latest version if there are several.
//    My ilasm.exe is in directory -
//        C:\WINDOWS\Microsoft.NET\Framework\v4.0.30319
//
package CP.MsilAsm;

public class MsilAsm {

    public static void Init() {
	    // empty
    }

    /**
     *  Copies the characters from the stream 
     *  <code>strm</code> to System.out, with every 
     *  line prefixed by the string "#ilasm".
     *  @throws IOException
     */
    static void CopyStream(java.io.InputStream strm) 
	    throws java.io.IOException {
        int chr;
        String prefix = "#ilasm: ";
        System.out.print(prefix);
        while ((chr = strm.read()) != -1) {
            System.out.print((char)chr);
            if (chr == (int)'\n')
                System.out.print(prefix);
        }
        System.out.println();
    }

    public static void Assemble(char[] fil, 
                                char[] opt, 
                                boolean main) {
	    // empty
    }

    public static int DoAsm(char[] fil,   // Command name 
                            char[] opt,   // Options, usually /debug
                            boolean main, // true ==> EXE, else DLL
                            boolean vrbs) // true ==> -verbose flag set
    {
        int rslt = 0;
        java.util.ArrayList<String> cmdList = new java.util.ArrayList<String>();
	//
	// Arg[0] is the command name
	//
        cmdList.add("ilasm");
        cmdList.add(vrbs ? "/nologo" : "/quiet");
        cmdList.add(main ? "/exe" : "/dll");
	//
	//  Now the rest of the user-supplied args
	//
        String[] args = new String(opt).trim().split(" ");
        for (int i = 0; i < args.length; i++)
            if (args[i].length() > 0)
                cmdList.add(args[i]);
        //
	// Now add the IL source file name
	//
        String fName = new String(fil).trim();
        cmdList.add(fName + ".il");

        String sffx = (main ? ".exe" : ".dll");
        // 
        if (vrbs) {
            System.out.print("#gpcp: MsilAsm spawning -\n       ");
            for (int i = 0; i < cmdList.size(); i++)
                 System.out.print(cmdList.get(i) + " ");
            System.out.println();
        }
        //
        try {
            Process proc = 
                new ProcessBuilder(cmdList).redirectErrorStream(true).start();
	    // Write output of process to System.out
            CopyStream(proc.getInputStream());
	    // wait for process to exit.
            rslt = proc.waitFor();
        }
        catch (Exception exc) {
            rslt = 2;
        }
        if (rslt == 0)
	    System.out.println("#gpcp: Created " + fName + (main ? ".exe" : ".dll"));
        else
	    System.out.println("#gpcp: ILASM FAILED");
        return rslt;
    }
}

File: sources/gpcp/csharp/MsilAsm.cs
// (* ========================================================= *)
// (**	Interface to the ILASM Byte-code assembler.		*)
// (*	K John Gough, 10th June 1999				*)
// (*	Modifications:						*)
// (*		Version for GPCP V0.3 April 2000 (kjg)		*)
// (* ========================================================= *)
// (*	The real code is in MsilAsm.cs				*)	
// (* ========================================================= *)
//
//MODULE MsilAsm;
//
//  PROCEDURE Init*(); BEGIN END Init;
//
//  PROCEDURE Assemble*(IN fil, opt : ARRAY OF CHAR; main : BOOLEAN); 
//  BEGIN END Assemble;
//
//END MsilAsm.
//
// Compile with :
//
//  $ csc /t:library /debug MsilAsm.cs
// 

using System;
using System.Text;
using System.Diagnostics;

namespace MsilAsm {

public class MsilAsm {

    private static Process asm = null;


    public static string GetDotNetRuntimeInstallDirectory() 
    {
        // Get the path to mscorlib.dll
        string s = typeof(object).Module.FullyQualifiedName;
        // Remove the file part to get the directory
        return System.IO.Directory.GetParent(s).ToString();
    }

        public static void Init() {
            if (asm == null) {
                asm = new Process();
                System.String frameworkDir = GetDotNetRuntimeInstallDirectory();
                //System.String frameworkDir = Environment.GetEnvironmentVariable("NET40", EnvironmentVariableTarget.User);
                asm.StartInfo.FileName = frameworkDir
                  + (frameworkDir[0] == '/' ?
                        "/../../../Commands/" : "\\") + "ilasm";

                asm.StartInfo.CreateNoWindow = true;
                asm.StartInfo.UseShellExecute = false;
            }
        }

        private static int CPlen(char[] arr) {
            int len = arr.Length;
            for (int ix = 0; ix < len; ix++)
                if (arr[ix] == '\0')
                    return ix;
            return len;
        }

        public static void Assemble(char[] fil, char[] opt, bool hasMain) {
            int retCode;
            System.String optNm;
            System.String suffx;
            System.String fName = new String(fil, 0, CPlen(fil));
            if (hasMain) {
                optNm = "/exe ";
                suffx = ".exe";
            }
            else {
                optNm = "/dll ";
                suffx = ".dll";
            }
            optNm = optNm + new String(opt, 0, CPlen(opt)) + ' ';
            asm.StartInfo.Arguments = optNm + "/nologo /quiet " + fName + ".il";
            asm.Start();
            asm.WaitForExit();
            retCode = asm.ExitCode;
            if (retCode != 0)
                System.Console.WriteLine("#gpcp: ilasm FAILED " + retCode);
            else
                System.Console.WriteLine("#gpcp: created " + fName + suffx);
        }

        public static void DoAsm(char[] fil, char[] opt,
                    bool hasMain,
                    bool verbose,
                    ref int rslt) {
            System.String optNm;
            System.String suffx;
            System.String fName = new String(fil, 0, CPlen(fil));

            if (hasMain) {
                optNm = "/exe ";
                suffx = ".exe";
            }
            else {
                optNm = "/dll ";
                suffx = ".dll";
            }
            optNm = optNm + new String(opt, 0, CPlen(opt)) + ' ';
            if (verbose) {
                asm.StartInfo.Arguments = optNm + "/nologo " + fName + ".il";
                System.Console.WriteLine("#gpcp: Calling " + asm.StartInfo.FileName + ' ' + asm.StartInfo.Arguments);
            }
            else {
                asm.StartInfo.Arguments = optNm + "/nologo /quiet " + fName + ".il";
            }
            asm.Start();
            asm.WaitForExit();
            rslt = asm.ExitCode;
            if (rslt == 0)
                System.Console.WriteLine("#gpcp: Created " + fName + suffx);
        }


        public static void Assemble(char[] fil, bool hasMain) {
        char[] opt = {'/', 'd', 'e', 'b', 'u', 'g', '\0' };
        Assemble(fil, opt, hasMain);
    }

  }
}
